_source/_posts/2021-12-06-spring-data-elasticsearch.md

```diff
@@ -0,0 +1,369 @@
+            client-secret: $SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_SECRET\" >> docker-compose/central-server-config/application.yml
+```
+
+
+For the experiment in this tutorial, configure the Kibana interface for Elasticsearch, which will allow visualizations of the Elasticsearch data.
+Edit `docker-compose/docker-compose.yml` and add the Kibana service like this:
+
+```yml
+services:
+  ...  
+  blog-kibana:
+     image: docker.elastic.co/kibana/kibana:7.15.2
+     ports:
+       - 5601:5601
+     environment:
+       ELASTICSEARCH_URL: http://blog-elasticsearch:9200
+       ELASTICSEARCH_HOSTS: '["http://blog-elasticsearch:9200"]'
+```
+**NOTE**: The Kibana version must match the version of Elasticsearch that's used by the `blog-elasticsearch` image.
+
+Also add the following additional configuration to the `blog-elasticsearch` service, to expose the ports to the local host for sending APIs requests:
+
+```yml
+services:
+  blog-elasticsearch:
+    ...
+    ports:
+      - 9200:9200
+```
+Remove the Keycloak service at the bottom, as Okta will be used as Identity Provider.
+
+Create each applications' container image:
+
+```shell
+cd blog
+./mvnw -DskipTests -ntp -Pprod verify jib:dockerBuild
+```
+
+```shell
+cd gateway
+./mvnw -DskipTests -ntp -Pprod verify jib:dockerBuild
+```
+
+Run all your images with Docker Compose:
+
+```shell
+cd ../docker-compose
+docker compose up
+```
+ Access the JHipster Registry at `http://localhost:8761` and sign in with your Okta credentials. When you see all services up and green, go to `http://localhost:8080` and sign in with your Okta account.
+
+ {% img blog/spring-data-elasticsearch/okta-signin.png alt:"Okta Sign In Form" width:"400" %}{: .center-image }
+
+
+ **IMPORTANT NOTE**: There is a potential issue observed for the blog service, as [Kubernetes probes](https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.kubernetes-probes) are enabled by default, the blog application might change its state to `OUT_OF_SERVICE` during start up and Eureka discards the following heartbeats with state `UP`. In that case, disable Kubernetes probes in the `docker-compose.yml` with the following environment variable:
+
+```yml
+services:
+  blog:
+    image: blog
+    environment:
+      ...
+      - MANAGEMENT_ENDPOINT_HEALTH_PROBES_ENABLED=false
+```
+Restart your microservices stack using Ctrl+C and run `docker compose up` again.
+
+## Inspecting Elasticsearch index mapping
+
+During start-up, Spring Data Elasticsearch will create the index for the entities annotated with `@Document`, deriving the mappings from the entity's annotations. But for the properties, if the field type is not specified, it defaults to FieldType.Auto. This means, that no mapping entry is written for the property and that Elasticsearch will add a mapping entry dynamically when the first data for this property is stored.
```

Comment:
```suggestion
During startup, Spring Data Elasticsearch will create the index for the entities annotated with `@Document`, deriving the mappings from the entity's annotations. But for the properties, if the field type is not specified, it will default to FieldType.Auto. This means that no mapping entry is written for the property and that Elasticsearch will add a mapping entry dynamically when the first data for this property is stored.
```
style edits and tense consistency